### ISSUE TITLE
Calling consumer.Close takes ~4s to complete rd_kafka_destroy

### ISSUE THREAD

**Issue by JeanMertz** [confluentinc/confluent-kafka-go#131]
This isn't a bug report per-se, but I did notice that closing the consumer can take up to four seconds, running against a local single-broker Kafka cluster with no messages being sent or delivered.

Doing some testing, it seems the delay occurs here:

https://github.com/confluentinc/confluent-kafka-go/blob/da30423f069c3209c14c0fac76b7474bcb24faa1/kafka/consumer.go#L345

This isn't really an issue in production usage, but I was writing some integration tests, and opening/closing tens of consumers over several tests does make things considerably slow.

Obviously, I can work around this issue by using a global consumer for all the integration tests, but I was wondering if this is expected behaviour, or that there is something that can be tweaked to improve the closing duration.

I'm using the latest (HEAD) version of both `librdkafka` and `confluent-kafka-go`.

**Comment by edenhill** [comment]
This is usually caused by the client waiting for an outstanding protocol request to finish.

Can you reproduce this with `"debug": "protocol,broker,cgrp"` to let us figure out what is taking so long?

**Comment by JeanMertz** [comment]
Here's the log output, this output happens between two of my own log outputs I put above and below the before-mentioned function call:

```
2018/02/13 17:37:18 kafka: starting to close #4
%7|1518539838.995|DESTROY|rdkafka#consumer-1| [thrd:app]: Terminating instance
%7|1518539838.995|DESTROY|rdkafka#consumer-1| [thrd:main]: Destroy internal
%7|1518539838.995|DESTROY|rdkafka#consumer-1| [thrd:main]: Removing all topics
%7|1518539838.995|TERM|rdkafka#consumer-1| [thrd::0/internal]: :0/internal: Received TERMINATE op in state UP: 1 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs
%7|1518539838.995|TERM|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received TERMINATE op in state UP: 3 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 2 waitresps, 0 retrybufs
%7|1518539839.101|TERMINATE|rdkafka#consumer-1| [thrd::0/internal]: :0/internal: Handle is terminating: failed 0 request(s) in retry+outbuf
%7|1518539839.101|TERMINATE|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Handle is terminating: failed 0 request(s) in retry+outbuf
%7|1518539839.101|BROKERFAIL|rdkafka#consumer-1| [thrd::0/internal]: :0/internal: failed: err: Local: Broker handle destroyed: (errno: Operation timed out)
%7|1518539839.101|BROKERFAIL|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: failed: err: Local: Broker handle destroyed: (errno: Operation timed out)
%7|1518539839.101|STATE|rdkafka#consumer-1| [thrd::0/internal]: :0/internal: Broker changed state UP -> DOWN
%7|1518539839.101|STATE|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Broker changed state UP -> DOWN
%7|1518539839.101|TERM|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Received TERMINATE op in state DOWN: 1 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs
%7|1518539841.928|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received JoinGroupResponse (v0, 180 bytes, CorrId 4, rtt 3030.77ms)
%7|1518539841.928|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received LeaveGroupResponse (v0, 2 bytes, CorrId 5, rtt 2933.07ms)
%7|1518539842.029|TERMINATE|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Handle is terminating: failed 0 request(s) in retry+outbuf
%7|1518539842.029|BROKERFAIL|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: failed: err: Local: Broker handle destroyed: (errno: Operation timed out)
%7|1518539842.029|STATE|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Broker changed state UP -> DOWN
%7|1518539842.030|MEMBERID|rdkafka#consumer-1| [thrd:app]: Group "test": updating member id "" -> "(not-set)"
2018/02/13 17:37:22 kafka: starting to close #5
```

Looks like the biggest delay happens here:

```
%7|1518539839.101|TERM|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Received TERMINATE op in state DOWN: 1 refcnts, 0 toppar(s), 0 active toppar(s), 0 outbufs, 0 waitresps, 0 retrybufs
%7|1518539841.928|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received JoinGroupResponse (v0, 180 bytes, CorrId 4, rtt 3030.77ms)
```

**Comment by edenhill** [comment]
It looks like the group is rebalancing just before you terminate your application, causing a JoinGroupRequest to block (on the broker) for 3s:
`Received JoinGroupResponse (v0, 180 bytes, CorrId 4, rtt 3030.77ms)`

We could probably optimize this case and forget about the outstanding JoinGroupRequest to speed up termination, but it might not help much since there is a LeaveGroupRequest sent after the JoinGroupRequest and it is being held back (by the broker) until the JoinGroupRequest has been served.

**Comment by JeanMertz** [comment]
I see. I'm not quite sure I understand why this happens though:

* it's not random, it happens 100% of the time when I call `consumer.Close()`
* I disabled all but one tests, and this Kafka instance isn't serving any other clients when I run the test, so why would it trigger a `JoinGroupResponse` when I'm actually in the process of leaving the group after calling close?

**Comment by edenhill** [comment]
Can you show me some of your code ,in particular the teardown?

**Comment by JeanMertz** [comment]
I can try to show the relevant parts.

Here's the test itself (I removed everything but creating and removing the client):

Actually, looking at _all_ the debug output, there's only _one_ `JoinGroupResponse` log line, so perhaps this is the one that is initiated after connecting the consumer? Either way, here's the [full log of the entire run in debug mode](https://gist.github.com/JeanMertz/612f4fb7b8625c082c7b9b435a27fd10).

(note that this uses a wrapper client, so the API here isn't the one used by this library, as you can see)

```golang
func TestConsumer(t *testing.T) {
	_, closer := newConsumer(t, "127.0.0.1:9092")
	defer closer()
}

func newConsumer(tb testing.TB, b string) (stream.Consumer, func()) {
	client, err := kafkaclient.New()
	if err != nil {
		tb.Fatalf("Unexpected error: %v", err)
	}

	options := func(c *streamconfig.Consumer) {
		c.Kafka.Brokers = []string{b}
	}

	consumer, err := client.NewConsumer(options)
	if err != nil {
		tb.Fatalf("Unexpected error: %v", err)
	}

	fn := func() {
		err = consumer.Close()
		if err != nil {
			tb.Fatalf("Unexpected error: %v", err)
		}
	}

	return consumer, fn
}
```

and here's where the consumer is instantiated:

```golang
func newConsumer(c *Client, options []func(*streamconfig.Consumer)) (*Consumer, error) {
	config, err := streamconfig.NewConsumer(c.rawConfig, options...)
	if err != nil {
		return nil, err
	}

	brokers := strings.Join(config.Kafka.Brokers, ",")

	kafkaconfig := &kafka.ConfigMap{
		"metadata.broker.list":     brokers,
		"group.id":                 "test",
		"session.timeout.ms":       6000,
		"go.events.channel.enable": true,
		"debug":                    "protocol,broker,cgrp",
		"go.application.rebalance.enable": true,
		"default.topic.config": kafka.ConfigMap{
			"auto.offset.reset": "earliest",
		},
	}

	kafkaconsumer, err := kafka.NewConsumer(kafkaconfig)
	if err != nil {
		return nil, err
	}

	err = kafkaconsumer.SubscribeTopics([]string{"topics..."}, nil)
	if err != nil {
		return nil, err
	}

	consumer := &Consumer{
		config:    config.Kafka,
		rawConfig: config,
		kafka:     kafkaconsumer,
		messages:  make(chan streammsg.Message),
		quit:      make(chan bool),
	}

	return consumer, nil
}
```

The function that handles the consuming of messages:

```golang
func (c *Consumer) consume() {
	defer func() {
		close(c.messages)
		c.wg.Done()
	}()

	for {
		select {
		case <-c.quit:
			return
		case event := <-c.kafka.Events():
			switch e := event.(type) {

			// If we received an `AssignedPartitions` event, we need to make sure we
			// assign the currently running consumer to the right partitions.
			case kafka.AssignedPartitions:
				c.kafka.Assign(e.Partitions)

			// If we received an `RevokedPartitions` event, we need to revoke this
			// consumer from all partitions. This means this consumer won't pick up
			// any work anymore, until a new `AssignedPartitions` event is handled.
			case kafka.RevokedPartitions:
				c.kafka.Unassign()

			// If we receive an error, something happened on Kafka's side. We don't
			// know what happened or if we can recover gracefully, so we instead
			// terminate the running process.
			case kafka.Error:
				c.config.Logger.Fatal(
					"Received Kafka consumer error. Exiting.",
					zap.Error(e),
				)

			// On receiving a Kafka message, we process the received message and
			// prepare it for delivery to the consumer of the consumer.messages
			// channel.
			case *kafka.Message:
				msg := c.newMessage(e)

				// Once the message has been prepared, we offer it to the consumer of
				// the messages channel. Since this is a blocking channel, we also
				// listen for the quit signal, and stop delivering new messages
				// accordingly.
				select {
				case c.messages <- msg:
				case <-c.quit:
					return
				}
			}
		}
	}
}
```

The close function:

```golang
func (c *Consumer) Close() error {
	err := c.kafka.Close()
	if err != nil {
		return err
	}

	c.quit <- true

	// Wait until the WaitGroup counter is zero. This makes sure we block the
	// close call until the reader has been closed, to prevent reading errors.
	c.wg.Wait()

	return nil
}
```

And finally the `NewConsumer` function as called from the tests:

```golang
func (c *Client) NewConsumer(options ...func(*streamconfig.Consumer)) (stream.Consumer, error) {
	consumer, err := newConsumer(c, options)
	if err != nil {
		return nil, err
	}

	consumer.wg.Add(1)
	go consumer.consume()

	return consumer, nil
}
```

**Comment by edenhill** [comment]
Thank you.

If I'm reading your code correctly the TestConsumer() will fire off the deferred close function straight away, so it is likely that the initial JoinGroupRequest (trigger by SubscribeTopics()) is indeed in-flight when consumer.Close() is called, causing the LeaveGroup to be blocked until the JoinGroupResponse is received.

To solve this we could check if there is a need to actually wait for the LeaveGroup to succeed before exiting, e.g., by looking at outstanding commits and the assignment count.

On the other hand, this seems like a flaw in the test itself: there's code to handle AssignedPartitions and Message events, but they are unlikely to ever kick in since the Close is immediate. If the test was to wait for the assignment then the close() would probably be immediate.

**Comment by JeanMertz** [comment]
@edenhill thank you for that.

Yes, indeed, there's nothing happening between starting and closing the connection. This is something that is still to be implemented, but I came across this 3 second run _even without any actual action_, so I figured that whatever simple test I add, it will always take 3 seconds _per test case_ (assuming I open and close a connection for each case), that's why I was wondering if this is to be expected or not.

Specifically, when using Sarama before, it did not have this 3 seconds run per test-case, this is obviously not a comparison that can be made, as it might be that Sarama is handling things differently, or incorrectly, but if I understand you correctly, the three seconds duration (with only opening and closing a connection) is as expected, correct?

**Comment by edenhill** [comment]
Yeah, this is expected behaviour: the client attempts to terminate cleanly by leaving the group so that remaining members of the group can rebalance quickly, and that LeaveGroup is blocked by a JoinGroup on the broker.

**Comment by JeanMertz** [comment]
Alright thanks. I just didn't expect those actions to take ~4s on Kafka's side, especially on a single-broker cluster that runs locally, and has no other activities. But that's obviously not the fault of this client, so I'll close this issue.

Thanks again for looking into this and explaining it to me 👍

**Comment by JeanMertz** [comment]
One more thing, I was searching for `JoinGroupRequest` through issues on this repository, and I found this posted log:

https://github.com/confluentinc/confluent-kafka-go/issues/71#issuecomment-314090988:

```
%7|1499689330.547|SEND|rdkafka#consumer-288| [thrd:kc1.docker:9092/bootstrap]: kc1.docker:9092/1: Sent JoinGroupRequest (v0, 116 bytes @ 0, CorrId 3)
%7|1499689330.547|BROADCAST|rdkafka#consumer-288| [thrd:kc1.docker:9092/bootstrap]: Broadcasting state change
%7|1499689330.548|RECV|rdkafka#consumer-288| [thrd:kc1.docker:9092/bootstrap]: kc1.docker:9092/1: Received JoinGroupResponse (v0, 179 bytes, CorrId 3, rtt 1.29ms)
```

As you can see, the response takes a total of 1.29ms.

Comparing that with my test, for which I've removed the close call and only do the join, it takes 4000ms to receive a response from Kafka for joining a group.

Using this test:

```golang
func TestConsumer(t *testing.T) {
	_, _, _ = newConsumer(t, "127.0.0.1:9092")
	time.Sleep(5 * time.Second)
}
```

I get this debug output:

```
%7|1518564341.331|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent JoinGroupRequest (v0, 106 bytes @ 0, CorrId 4)
%7|1518564344.372|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received JoinGroupResponse (v0, 175 bytes, CorrId 4, rtt 3041.53ms)
```

<details>
 <summary>See full log output</summary>

```
%7|1518564341.327|MEMBERID|rdkafka#consumer-1| [thrd:app]: Group "test": updating member id "(not-set)" -> ""
%7|1518564341.327|BRKREASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test" management reassigned from broker (none) to :0/internal
%7|1518564341.327|CGRPSTATE|rdkafka#consumer-1| [thrd:main]: Group "test" changed state init -> wait-broker (v1, join-state init)
%7|1518564341.327|BRKMAIN|rdkafka#consumer-1| [thrd::0/internal]: :0/internal: Enter main broker thread
%7|1518564341.327|BROKER|rdkafka#consumer-1| [thrd:app]: 127.0.0.1:9092/bootstrap: Added new broker with NodeId -1
%7|1518564341.327|STATE|rdkafka#consumer-1| [thrd::0/internal]: :0/internal: Broker changed state INIT -> UP
%7|1518564341.327|BRKASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test" management assigned to broker :0/internal
%7|1518564341.327|BRKMAIN|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Enter main broker thread
%7|1518564341.327|CONNECT|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: broker in state INIT connecting
%7|1518564341.327|CGRPOP|rdkafka#consumer-1| [thrd:main]: Group "test" received op SUBSCRIBE (v0) in state wait-broker (join state init, v1 vs 0)
%7|1518564341.327|SUBSCRIBE|rdkafka#consumer-1| [thrd:main]: Group "test": subscribe to new subscription of 1 topics (join state init)
%7|1518564341.327|UNSUBSCRIBE|rdkafka#consumer-1| [thrd:main]: Group "test": unsubscribe from current unset subscription of 0 topics (leave group=no, join state init, v1)
%7|1518564341.327|GRPLEADER|rdkafka#consumer-1| [thrd:main]: Group "test": resetting group leader info: unsubscribe
%7|1518564341.327|CGRPJOINSTATE|rdkafka#consumer-1| [thrd:main]: Group "test" changed join state init -> wait-unassign (v1, state wait-broker)
%7|1518564341.327|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test": unassign done in state wait-broker (join state wait-unassign): without new assignment: unassign (no previous assignment)
%7|1518564341.327|CGRPJOINSTATE|rdkafka#consumer-1| [thrd:main]: Group "test" changed join state wait-unassign -> init (v1, state wait-broker)
%7|1518564341.327|CONNECT|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Connecting to ipv4#127.0.0.1:9092 (plaintext) with socket 7
%7|1518564341.327|CGRPQUERY|rdkafka#consumer-1| [thrd:main]: Group "test": no broker available for coordinator query: intervaled in state wait-broker
%7|1518564341.327|STATE|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Broker changed state INIT -> CONNECT
%7|1518564341.327|CONNECT|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Connected to ipv4#127.0.0.1:9092
%7|1518564341.327|CONNECTED|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Connected (#1)
%7|1518564341.327|FEATURE|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Updated enabled protocol features +ApiVersion to ApiVersion
%7|1518564341.327|STATE|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Broker changed state CONNECT -> APIVERSION_QUERY
%7|1518564341.327|SEND|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Sent ApiVersionRequest (v0, 25 bytes @ 0, CorrId 1)
%7|1518564341.328|RECV|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Received ApiVersionResponse (v0, 234 bytes, CorrId 1, rtt 0.68ms)
%7|1518564341.328|FEATURE|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Updated enabled protocol features to MsgVer1,ApiVersion,BrokerBalancedConsumer,ThrottleTime,Sasl,SaslHandshake,BrokerGroupCoordinator,LZ4,OffsetTime,MsgVer2
%7|1518564341.328|STATE|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Broker changed state APIVERSION_QUERY -> UP
%7|1518564341.328|SEND|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Sent MetadataRequest (v2, 25 bytes @ 0, CorrId 2)
%7|1518564341.328|RECV|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Received MetadataResponse (v2, 61 bytes, CorrId 2, rtt 0.33ms)
%7|1518564341.329|BROKER|rdkafka#consumer-1| [thrd:main]: 192.168.1.152:9092/0: Added new broker with NodeId 0
%7|1518564341.329|CLUSTERID|rdkafka#consumer-1| [thrd:main]: 127.0.0.1:9092/bootstrap: ClusterId update "" -> "3f6o9SRxTcqNOe7MvVbucQ"
%7|1518564341.329|SUBSCRIPTION|rdkafka#consumer-1| [thrd:main]: Group "test": no topics in metadata matched subscription
%7|1518564341.329|BRKREASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test" management reassigned from broker :0/internal to 127.0.0.1:9092/bootstrap
%7|1518564341.329|BRKUNASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test" management unassigned from broker handle :0/internal
%7|1518564341.329|BRKASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test" management assigned to broker 127.0.0.1:9092/bootstrap
%7|1518564341.329|CGRPQUERY|rdkafka#consumer-1| [thrd:main]: 127.0.0.1:9092/bootstrap: Group "test": querying for coordinator: intervaled in state wait-broker
%7|1518564341.329|BRKMAIN|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Enter main broker thread
%7|1518564341.329|CONNECT|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: broker in state INIT connecting
%7|1518564341.329|SEND|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Sent GroupCoordinatorRequest (v0, 27 bytes @ 0, CorrId 3)
%7|1518564341.329|CONNECT|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Connecting to ipv4#192.168.1.152:9092 (plaintext) with socket 10
%7|1518564341.329|STATE|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Broker changed state INIT -> CONNECT
%7|1518564341.329|CONNECT|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Connected to ipv4#192.168.1.152:9092
%7|1518564341.329|CONNECTED|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Connected (#1)
%7|1518564341.329|FEATURE|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Updated enabled protocol features +ApiVersion to ApiVersion
%7|1518564341.329|STATE|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Broker changed state CONNECT -> APIVERSION_QUERY
%7|1518564341.329|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent ApiVersionRequest (v0, 25 bytes @ 0, CorrId 1)
%7|1518564341.330|RECV|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Received GroupCoordinatorResponse (v0, 25 bytes, CorrId 3, rtt 0.96ms)
%7|1518564341.330|CGRPCOORD|rdkafka#consumer-1| [thrd:main]: 127.0.0.1:9092/bootstrap: Group "test" coordinator is 192.168.1.152:9092 id 0
%7|1518564341.330|CGRPCOORD|rdkafka#consumer-1| [thrd:main]: Group "test" changing coordinator -1 -> 0
%7|1518564341.330|BRKREASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test" management reassigned from broker 127.0.0.1:9092/bootstrap to 192.168.1.152:9092/0
%7|1518564341.330|BRKUNASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test" management unassigned from broker handle 127.0.0.1:9092/bootstrap
%7|1518564341.330|BRKASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test" management assigned to broker 192.168.1.152:9092/0
%7|1518564341.330|CGRPSTATE|rdkafka#consumer-1| [thrd:main]: Group "test" changed state wait-broker -> wait-broker-transport (v1, join-state init)
%7|1518564341.330|CGRPQUERY|rdkafka#consumer-1| [thrd:main]: 127.0.0.1:9092/bootstrap: Group "test": querying for coordinator: intervaled in state wait-broker-transport
%7|1518564341.330|SEND|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Sent GroupCoordinatorRequest (v0, 27 bytes @ 0, CorrId 4)
%7|1518564341.330|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received ApiVersionResponse (v0, 234 bytes, CorrId 1, rtt 0.80ms)
%7|1518564341.330|FEATURE|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Updated enabled protocol features to MsgVer1,ApiVersion,BrokerBalancedConsumer,ThrottleTime,Sasl,SaslHandshake,BrokerGroupCoordinator,LZ4,OffsetTime,MsgVer2
%7|1518564341.330|STATE|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Broker changed state APIVERSION_QUERY -> UP
%7|1518564341.330|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent MetadataRequest (v2, 25 bytes @ 0, CorrId 2)
%7|1518564341.330|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received MetadataResponse (v2, 61 bytes, CorrId 2, rtt 0.25ms)
%7|1518564341.330|SUBSCRIPTION|rdkafka#consumer-1| [thrd:main]: Group "test": no topics in metadata matched subscription
%7|1518564341.330|CGRPSTATE|rdkafka#consumer-1| [thrd:main]: Group "test" changed state wait-broker-transport -> up (v1, join-state init)
%7|1518564341.330|JOIN|rdkafka#consumer-1| [thrd:main]: Group "test": join with 0 (1) subscribed topic(s)
%7|1518564341.330|CGRPMETADATA|rdkafka#consumer-1| [thrd:main]: consumer join: metadata for subscription only available for 0/1 topics (-1ms old)
%7|1518564341.330|RECV|rdkafka#consumer-1| [thrd:127.0.0.1:9092/bootstrap]: 127.0.0.1:9092/bootstrap: Received GroupCoordinatorResponse (v0, 25 bytes, CorrId 4, rtt 0.44ms)
%7|1518564341.330|JOIN|rdkafka#consumer-1| [thrd:main]: Group "test": postponing join until up-to-date metadata is available
%7|1518564341.330|CGRPCOORD|rdkafka#consumer-1| [thrd:main]: 127.0.0.1:9092/bootstrap: Group "test" coordinator is 192.168.1.152:9092 id 0
%7|1518564341.330|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent MetadataRequest (v2, 31 bytes @ 0, CorrId 3)
%7|1518564341.331|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received MetadataResponse (v2, 100 bytes, CorrId 3, rtt 0.32ms)
%7|1518564341.331|SUBSCRIPTION|rdkafka#consumer-1| [thrd:main]: Group "test": effective subscription list changed from 0 to 1 topic(s):
%7|1518564341.331|SUBSCRIPTION|rdkafka#consumer-1| [thrd:main]:  Topic test with 1 partition(s)
%7|1518564341.331|REJOIN|rdkafka#consumer-1| [thrd:main]: Group "test": subscription updated from metadata change: rejoining group
%7|1518564341.331|GRPLEADER|rdkafka#consumer-1| [thrd:main]: Group "test": resetting group leader info: Group rejoin
%7|1518564341.331|REJOIN|rdkafka#consumer-1| [thrd:main]: Group "test" rejoining in join-state init without an assignment
%7|1518564341.331|JOIN|rdkafka#consumer-1| [thrd:main]: Group "test": join with 1 (1) subscribed topic(s)
%7|1518564341.331|CGRPMETADATA|rdkafka#consumer-1| [thrd:main]: consumer join: metadata for subscription is up to date (0ms old)
%7|1518564341.331|CGRPJOINSTATE|rdkafka#consumer-1| [thrd:main]: Group "test" changed join state init -> wait-join (v1, state up)
%7|1518564341.331|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent JoinGroupRequest (v0, 106 bytes @ 0, CorrId 4)
%7|1518564344.372|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received JoinGroupResponse (v0, 175 bytes, CorrId 4, rtt 3041.53ms)
%7|1518564344.372|JOINGROUP|rdkafka#consumer-1| [thrd:main]: JoinGroup response: GenerationId 29, Protocol range, LeaderId rdkafka-5795a459-58d7-42e7-a54f-4df7d48e5916 (me), my MemberId rdkafka-5795a459-58d7-42e7-a54f-4df7d48e5916, 1 members in group: (no error)
%7|1518564344.372|MEMBERID|rdkafka#consumer-1| [thrd:main]: Group "test": updating member id "" -> "rdkafka-5795a459-58d7-42e7-a54f-4df7d48e5916"
%7|1518564344.372|JOINGROUP|rdkafka#consumer-1| [thrd:main]: Elected leader for group "test" with 1 member(s)
%7|1518564344.372|GRPLEADER|rdkafka#consumer-1| [thrd:main]: Group "test": resetting group leader info: JoinGroup response clean-up
%7|1518564344.372|CGRPJOINSTATE|rdkafka#consumer-1| [thrd:main]: Group "test" changed join state wait-join -> wait-metadata (v1, state up)
%7|1518564344.372|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent MetadataRequest (v2, 31 bytes @ 0, CorrId 5)
%7|1518564344.373|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received MetadataResponse (v2, 100 bytes, CorrId 5, rtt 0.60ms)
%7|1518564344.373|ASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test" running range assignment for 1 member(s):
%7|1518564344.373|ASSIGN|rdkafka#consumer-1| [thrd:main]:  Member "rdkafka-5795a459-58d7-42e7-a54f-4df7d48e5916" (me) with 1 subscription(s):
%7|1518564344.373|ASSIGN|rdkafka#consumer-1| [thrd:main]:   test [-1]
%7|1518564344.373|ASSIGN|rdkafka#consumer-1| [thrd:main]: range: Topic test with 1 partition(s) and 1 subscribing member(s)
%7|1518564344.373|ASSIGN|rdkafka#consumer-1| [thrd:main]: range: Member "rdkafka-5795a459-58d7-42e7-a54f-4df7d48e5916": assigned topic test partitions 0..0
%7|1518564344.373|ASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test" range assignment for 1 member(s) finished in 0.031ms:
%7|1518564344.373|ASSIGN|rdkafka#consumer-1| [thrd:main]:  Member "rdkafka-5795a459-58d7-42e7-a54f-4df7d48e5916" (me) assigned 1 partition(s):
%7|1518564344.373|ASSIGN|rdkafka#consumer-1| [thrd:main]:   test [0]
%7|1518564344.373|ASSIGNOR|rdkafka#consumer-1| [thrd:main]: Group "test": "range" assignor run for 1 member(s)
%7|1518564344.373|CGRPJOINSTATE|rdkafka#consumer-1| [thrd:main]: Group "test" changed join state wait-metadata -> wait-sync (v1, state up)
%7|1518564344.373|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent SyncGroupRequest (v0, 155 bytes @ 0, CorrId 6)
%7|1518564344.375|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received SyncGroupResponse (v0, 30 bytes, CorrId 6, rtt 1.24ms)
%7|1518564344.375|SYNCGROUP|rdkafka#consumer-1| [thrd:main]: SyncGroup response: Success (24 bytes of MemberState data)
%7|1518564344.375|ASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test": delegating assign of 1 partition(s) to application rebalance callback on queue rd_kafka_cgrp_new: new assignment
%7|1518564344.375|CGRPJOINSTATE|rdkafka#consumer-1| [thrd:main]: Group "test" changed join state wait-sync -> wait-assign-rebalance_cb (v1, state up)
%7|1518564344.375|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: 192.168.1.152:9092/0: Heartbeat for group "test" generation id 29
%7|1518564344.375|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent HeartbeatRequest (v0, 77 bytes @ 0, CorrId 7)
%7|1518564344.375|CGRPOP|rdkafka#consumer-1| [thrd:main]: Group "test" received op ASSIGN (v0) in state up (join state wait-assign-rebalance_cb, v1 vs 0)
%7|1518564344.375|ASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test": new assignment of 1 partition(s) in join state wait-assign-rebalance_cb
%7|1518564344.375|BARRIER|rdkafka#consumer-1| [thrd:main]: Group "test": rd_kafka_cgrp_assign:2299: new version barrier v2
%7|1518564344.375|TOPBRK|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Topic test [0]: joining broker (rktp 0x7800880)
%7|1518564344.375|ASSIGN|rdkafka#consumer-1| [thrd:main]: Group "test": assigning 1 partition(s) in join state wait-assign-rebalance_cb
%7|1518564344.375|CGRPJOINSTATE|rdkafka#consumer-1| [thrd:main]: Group "test" changed join state wait-assign-rebalance_cb -> assigned (v2, state up)
%7|1518564344.375|BARRIER|rdkafka#consumer-1| [thrd:main]: Group "test": rd_kafka_cgrp_partitions_fetch_start0:1640: new version barrier v3
%7|1518564344.375|FETCHSTART|rdkafka#consumer-1| [thrd:main]: Group "test": starting fetchers for 1 assigned partition(s) in join-state assigned (usable_offsets=no, v3, line 2344)
%7|1518564344.375|FETCHSTART|rdkafka#consumer-1| [thrd:main]: List with 1 partition(s):
%7|1518564344.375|FETCHSTART|rdkafka#consumer-1| [thrd:main]:  test [0] offset INVALID
%7|1518564344.375|OFFSET|rdkafka#consumer-1| [thrd:main]: 192.168.1.152:9092/0: Fetch committed offsets for 1/1 partition(s)
%7|1518564344.375|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent OffsetFetchRequest (v1, 45 bytes @ 0, CorrId 8)
%7|1518564344.375|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received HeartbeatResponse (v0, 2 bytes, CorrId 7, rtt 0.38ms)
%7|1518564344.375|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received OffsetFetchResponse (v1, 30 bytes, CorrId 8, rtt 0.43ms)
%7|1518564344.375|OFFSETFETCH|rdkafka#consumer-1| [thrd:main]: List with 1 partition(s):
%7|1518564344.375|OFFSETFETCH|rdkafka#consumer-1| [thrd:main]:  test [0] offset INVALID
%7|1518564344.376|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic test [0]: setting default offset INVALID
%7|1518564344.376|BARRIER|rdkafka#consumer-1| [thrd:main]: Group "test": rd_kafka_cgrp_partitions_fetch_start0:1640: new version barrier v4
%7|1518564344.376|FETCHSTART|rdkafka#consumer-1| [thrd:main]: Group "test": starting fetchers for 1 assigned partition(s) in join-state assigned (usable_offsets=yes, v4, line 1572)
%7|1518564344.376|FETCHSTART|rdkafka#consumer-1| [thrd:main]: List with 1 partition(s):
%7|1518564344.376|FETCHSTART|rdkafka#consumer-1| [thrd:main]:  test [0] offset 6
%7|1518564344.376|CGRPJOINSTATE|rdkafka#consumer-1| [thrd:main]: Group "test" changed join state assigned -> started (v4, state up)
%7|1518564344.376|CGRPOP|rdkafka#consumer-1| [thrd:main]: Group "test" received op PARTITION_JOIN in state up (join state started, v4) for test [0]
%7|1518564344.376|PARTADD|rdkafka#consumer-1| [thrd:main]: Group "test": add test [0]
%7|1518564344.376|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 9)
%7|1518564344.376|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 206 bytes, CorrId 9, rtt 0.56ms)
%7|1518564344.376|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 10)
%7|1518564344.489|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 10, rtt 112.65ms)
%7|1518564344.489|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 11)
%7|1518564344.654|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 11, rtt 165.24ms)
%7|1518564344.654|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 12)
%7|1518564344.818|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 12, rtt 163.33ms)
%7|1518564344.818|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 13)
%7|1518564344.973|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 13, rtt 155.09ms)
%7|1518564344.973|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 14)
%7|1518564345.103|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 14, rtt 129.86ms)
%7|1518564345.103|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 15)
%7|1518564345.239|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 15, rtt 135.96ms)
%7|1518564345.239|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 16)
%7|1518564345.389|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 16, rtt 150.43ms)
%7|1518564345.389|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 17)
%7|1518564345.557|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 17, rtt 167.49ms)
%7|1518564345.557|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 18)
%7|1518564345.719|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 18, rtt 161.50ms)
%7|1518564345.719|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 19)
%7|1518564345.878|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 19, rtt 158.92ms)
%7|1518564345.878|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 20)
%7|1518564346.033|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 20, rtt 155.57ms)
%7|1518564346.033|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 21)
%7|1518564346.198|RECV|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Received FetchResponse (v4, 48 bytes, CorrId 21, rtt 164.76ms)
%7|1518564346.198|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent FetchRequest (v4, 68 bytes @ 0, CorrId 22)
%7|1518564346.327|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: 192.168.1.152:9092/0: Heartbeat for group "test" generation id 29
%7|1518564346.327|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent HeartbeatRequest (v0, 77 bytes @ 0, CorrId 23)
%7|1518564346.327|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic test [0]: stored offset 8, committed offset 6
%7|1518564346.327|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic test [0]: setting stored offset 8 for commit
%7|1518564346.327|SEND|rdkafka#consumer-1| [thrd:192.168.1.152:9092/0]: 192.168.1.152:9092/0: Sent OffsetCommitRequest (v1, 113 bytes @ 0, CorrId 24)
```
</details>

**Comment by edenhill** [comment]
The broker may wait up to session.timeout.ms for all members to (re)join on a rebalance caused by a JoinGroupRequest.

**Comment by JeanMertz** [comment]
Thank you for pointing me in the right direction. I reconfigured`zookeeper.connection.timeout.ms`,  `group.min.session.timeout.ms` and `group.max.session.timeout.ms`, set the new value from the client accordingly, but that unfortunately didn't have any effect.

However, I noticed the default of `group.initial.rebalance.delay.ms` was `3000`, which was around the delay of the response I was seeing (and, accordingly, the name obviously raises red flags regarding the issues I was seeing here). This indeed solved the problem and the delay is no longer present.

The documentation even [specifically mentions this flag](http://kafka.apache.org/documentation/#upgrade_1100_notable) in development/test situations:

> A new config, `group.initial.rebalance.delay.ms`, was introduced. This config specifies the time, in milliseconds, that the GroupCoordinator will delay the initial consumer rebalance. The rebalance will be further delayed by the value of `group.initial.rebalance.delay.ms` as new members join the group, up to a maximum of `max.poll.interval.ms`. The default value for this is 3 seconds. During development and testing it might be desirable to set this to 0 in order to not delay test execution time.

So this was clearly an instance of RTFM. Thank you for your patience and pointing me in the right direction 👍

**Comment by edenhill** [comment]
Glad you found it!
